### PR TITLE
Allow users to configure the Prometheus remote write queue

### PR DESCRIPTION
**Pull request by rakyll**
A follow up to https://github.com/open-telemetry/opentelemetry-collector/pull/2974.

### CODE REVIEW

**Comment by rakyll** [comment]
cc @odeke-em @Aneurysm9 @dashpole

File: exporter/prometheusremotewriteexporter/README.md
Comment (dashpole) [inline exporter/prometheusremotewriteexporter/README.md:40]: I don't see this in the PRW parameters?  Just to confirm, we are introducing something new here, right?  Is there a way for us to stick with the existing PRW configs?

Comment (rakyll) [reply exporter/prometheusremotewriteexporter/README.md:40]: This is new, PRW has aa setting for capacity but it represents the number of samples.

File: exporter/prometheusremotewriteexporter/factory.go
Comment (odeke-em) [inline exporter/prometheusremotewriteexporter/factory.go:64]: Thinks for fixing this, quite previously redundant.

**Review by odeke-em — APPROVED** [review]
LGTM, thank you @rakyll!

**Review by dashpole — APPROVED** [review]

**Comment by rakyll** [comment]
cc @alolita to label this PR.

**Comment by alolita** [comment]
@anuraaga @open-telemetry/collector-approvers can you please review so that we can merge this fix. We are blocked on this fix for other dependent prometheus issues.

**Comment by alolita** [comment]
@jrcamp @jpkrohling - can you please approve asap.

**Review by anuraaga — APPROVED** [review]

**Review by Aneurysm9 — APPROVED** [review]

**Comment by rakyll** [comment]
@jpkrohling can you assign someone?

File: exporter/prometheusremotewriteexporter/README.md
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/README.md:41]: The min_shards is not implemented correct? It does not do sharding or am I missing something?

Comment (rakyll) [reply exporter/prometheusremotewriteexporter/README.md:41]: Min shards controls remote_write fanout: https://github.com/open-telemetry/opentelemetry-collector/pull/3046/files#diff-2631d8a8855fdc08cf077c0ab99544147ca73eec572bc871d8fe717e826eda38R279-R298

From Prometheus:

```
      |-->  queue (shard_1)   --> remote endpoint
WAL --|-->  queue (shard_...) --> remote endpoint
      |-->  queue (shard_n)   --> remote endpoint
```

"Min shards configures the minimum number of shards used by Prometheus, and is the number of shards used when remote write starts. If remote write falls behind, Prometheus will automatically scale up the number of shards so most users do not have to adjust this parameter. However, increasing min shards will allow Prometheus to avoid falling behind at the beginning while calculating the required number of shards."

This is very similar to the fanout we have in the exporter. We haven't implemented max_shards yet to automatically increase the number of shards, but this PR is a good first step to give users a way to fine tune.

**Review by gramidt — APPROVED** [review]

File: exporter/prometheusremotewriteexporter/README.md
Comment (bogdandrutu) [reply exporter/prometheusremotewriteexporter/README.md:41]: I do understand what PRW min_shards does, but in this PR min_shards configures the concurrency, but does not split the incoming requests (a.k.a "sharding") to ensure that for one worker we always get the same shard (a.k.a I think we just add back the out of order problem?)

Probably I am missing something, but from the code in this PR this is what I am seeing, probably there are other internal things that happen, if that is the case please point me to that code.

**Comment by rakyll** [comment]
Can someone merge this PR? Not sure what else needed to be done. We need to implement follow up features and this PR has been blocking me for the last 15 days.

cc @bogdandrutu

**Comment by alolita** [comment]
@rakyll Please see @bogdandrutu 's comment above. He did review the PR to merge but wanted some clarification from you.

File: exporter/prometheusremotewriteexporter/README.md
Comment (rakyll) [reply exporter/prometheusremotewriteexporter/README.md:41]: batchTimeSeries shard the samples by timeseries. https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/prometheusremotewriteexporter/exporter.go#L273 Each request only contains samples from a particular timeseries already. This PR is only making something hardcoded a configurable setting.

**Comment by rakyll** [comment]
Sorry, just saw that. Addressed.

File: exporter/prometheusremotewriteexporter/README.md
Comment (rakyll) [reply exporter/prometheusremotewriteexporter/README.md:41]: Maybe, I'm misreading your comment here. If you are asking if we are implementing the shards and queues exactly the way Prometheus does, we don't. But this is something we can start with until we have time to replicate their queuing mechanism or reuse their code.

**Comment by alolita** [comment]
Thanks @rakyll 

@bogdandrutu please review and merge.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (carlosalberto) [inline exporter/prometheusremotewriteexporter/exporter.go:285]: If `requests` get a value less than `minShards`, then concurrencyLimit will indeed be less than such `minShards` value, no?

**Review by carlosalberto — APPROVED** [review]

File: exporter/prometheusremotewriteexporter/README.md
Comment (bogdandrutu) [reply exporter/prometheusremotewriteexporter/README.md:41]: >  If you are asking if we are implementing the shards and queues exactly the way Prometheus does, we don't. 

@rakyll this is what I was asking

If we don't do, I know consistency in terms of names is good, but maybe add a caveat that the full functionality is not implemented, and for the moment this only controls the number of workers.

Comment (rakyll) [reply exporter/prometheusremotewriteexporter/README.md:41]: > and for the moment this only controls the number of workers

This is what min_shards in Prometheus does.

min_shards controls the number of shards being concurrently sent to a remote write endpoint.  We create n requests sharded by timeseries name. Then, we create up to min_shards workers to send remote writes. This is not different from what Prom is doing with min_shards. 

We just don't autoscale the workers up to max_shards. This is ok, we will error if users want to bring their existing config with max_shard. 

If we want to call this setting something else like "workers", we will have to deprecate it once we have min_shards and max_shards implemented.

Comment (rakyll) [reply exporter/prometheusremotewriteexporter/README.md:41]: The difference I forgot to mention in the previous comment is Prom always has min_shards workers available. We create up to min_shards shards when we have metrics to export. The workers are not continuously running. This is only a deal when there is too much congestion on the "export" method. We should fix the mechanism here either now and as a follow up PR.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (rakyll) [reply exporter/prometheusremotewriteexporter/exporter.go:285]: True, this is slightly different than how Prom always have up to min_shards workers. See my other comment at https://github.com/open-telemetry/opentelemetry-collector/pull/3046#discussion_r633721031. I'm happy to make this looking more similar to Prometheus.

File: exporter/prometheusremotewriteexporter/README.md
Comment (rakyll) [reply exporter/prometheusremotewriteexporter/README.md]: Renamed min_shards to concurrency. PTAL.

**Comment by bogdandrutu** [comment]
Failing unittests https://github.com/open-telemetry/opentelemetry-collector/pull/3046/checks?check_run_id=2603529811

**Comment by rakyll** [comment]
@bogdandrutu rebased.

File: exporter/prometheusremotewriteexporter/config.go
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/config.go:58]: If this is not same as PRW, maybe use NumWorkers to be consistent with QueueRetry config. If it is same as PRW just let me know and we can ignore the comment.

**Review by bogdandrutu — APPROVED** [review]
One last comment. Thank you

File: exporter/prometheusremotewriteexporter/config.go
Comment (rakyll) [reply exporter/prometheusremotewriteexporter/config.go:58]: You mean  NumConsumer, right? I should probably rename the Size to QueueSize to keep things consistent.

Comment (rakyll) [reply exporter/prometheusremotewriteexporter/config.go]: PTAL.

Renamed the queue_config to remote_write_queue. Renamed concurrency to num_consumers and size to queue_size. This looks much better and is more honest.

**Review by bogdandrutu — APPROVED** [review]

**Comment by jpkrohling** [comment]
Sorry, just saw that I was mentioned here. The amount of notifications I get from this repo made me miss it :-/ In the future, if possible, ping me on CNCF Slack and I'll review or assign it to someone.

**Pull request merged**
